flowapi/flowapi/permissions.py

```diff
@@ -267,83 +138,61 @@ def schema_to_scopes(schema: dict) -> Iterable[str]:
+    #  "spatial_aggregate",
+    #  "locations:admin_1",
+    #  "locations:admin_3",
+    #  "event_dates:1990-02-01:1992-03-04"
+    #  "event_type:mds",
+    #  "event_type:sms",
+    #  "subscriber_subset"
+
+    # Boolean permissions:
+    # Check run
+    # Check read
+    # Check subscriber subset
+    # Check event types
+    # Check query tree
+    # Check dates
+    resolved_queries = ResolvingParser(spec_string=dumps(schema)).specification[
+        "components"
+    ]["schemas"]["FlowmachineQuerySchema"]
+    unique_scopes = []
+    for tl_query in resolved_queries["oneOf"]:
+        tl_query_name = tl_query["properties"]["query_kind"]["enum"][0]
+        print(f"Looking for {tl_query_name}")
+        query_list = grab_on_key_list(
+            tl_query,
+            ["properties", "query_kind", "enum", 0],
+        )
+        if query_list == []:
+            return []
+        scopes_generator = (tl_scope_string(tl_query, query) for query in query_list)
+        unique_scopes += list(set.union(*scopes_generator))
+    return sorted(unique_scopes)
 
 
-@functools.singledispatch
-def query_to_scope_list(tree, paths=None, keep=["aggregation_unit"]) -> str:
+def tl_scope_string(tl_query, query_string) -> set:
     """
-
-    Parameters
-    ----------
-    tree : list of dict or dict
-    paths : tuple
-    keep : list of str
-        List of fields to include in scope strings
-
-    Yields
-    ------
-    str
-        Scope string
-
-
+    Given a top level (aggregate) query and a sub_query, return the scopes triplet for that query.
+    This is composed of the top_level query, allowable admin level
+    :param tl_query:
     """
-    yield from ()
-
-
-@query_to_scope_list.register
-def _(tree: list, paths=None, keep=["aggregation_unit"]) -> str:
-    if paths is None:
-        paths = tuple()
-    for v in tree:
-        yield from query_to_scope_list(v, paths, keep=keep)
-
-
-@query_to_scope_list.register
-def _(tree: dict, paths=None, keep=["aggregation_unit"]) -> str:
-    if paths is None:
-        paths = tuple()
+    out = set()
+    tl_query_name = tl_query["properties"]["query_kind"]["enum"][0]
     try:
-        q_kind = tree["query_kind"]
-        paths = (*paths, q_kind)
+        agg_units = tl_query["properties"]["aggregation_unit"]["enum"]
     except KeyError:
-        pass
-
-    yielded_any = False
-    for k, v in sorted(tree.items(), key=lambda x: x[0]):
-        if k == "query_kind":
-            continue
-        if k in keep:
-            yield ".".join((*paths, k, v))
-            yielded_any = True
-        elif isinstance(v, (dict, list)):
-            for t in query_to_scope_list(v, (*paths, k), keep=keep):
-                yield t
-                yielded_any = True
-    if not yielded_any and "query_kind" in tree:
-        yield ".".join(paths)
+        logging.warning(
+            f"No aggregation unit options for {tl_query_name}; "
+            f"this should be fixed once PR 5278 is merged"
```

Comment:
There will always be some top-level queries without an aggregation unit, because not all aggregates are spatial aggregates (currently `histogram_aggregate` is the only non-spatial aggregate, I think). I think we want something like `agg_units = ["nonspatial"]` in that situation.